certbot/certbot/crypto_util.py

```diff
@@ -174,18 +185,39 @@ def import_csr_file(csrfile, data):
     return PEM, util.CSR(file=csrfile, data=data_pem, form="pem"), domains
 
 
-def make_key(bits):
-    """Generate PEM encoded RSA key.
+def make_key(bits=1024, key_type="rsa", elliptic_curve=None):
+    """Generate PEM encoded RSA|EC key.
 
-    :param int bits: Number of bits, at least 1024.
+    :param int bits: Number of bits if key_type=rsa. At least 1024 for RSA.
 
-    :returns: new RSA key in PEM form with specified number of bits
-    :rtype: str
+    :param str ec_curve: The elliptic curve to use.
 
+    :returns: new RSA or ECDSA key in PEM form with specified number of bits
+              or of type ec_curve when key_type ecdsa is used.
+    :rtype: str
     """
-    assert bits >= 1024  # XXX
-    key = crypto.PKey()
-    key.generate_key(crypto.TYPE_RSA, bits)
+    if key_type == 'rsa':
+        if bits < 1024 or bits % 1024 != 0:
```

Comment:
I know [boulder recently added this restriction](https://github.com/letsencrypt/boulder/pull/4839) b/c of [Debian Weak Keys](https://community.letsencrypt.org/t/issuing-for-common-rsa-key-sizes-only/133839). However, shouldn't certbot be able to generate RSA keys with modulo 8 in stead of 1024, to keep all options open? AFAIK certbot isn't a Let's Encrypt specific ACME client, correct? Other CA's might choose *not* to restrict RSA keys to modulo 1024.

Reply:
@osirisinferi @adferrand Perhaps it's actually better to do this restriction in a separate PR? I agree that 2048 would probably be a good lower limit for the key size. RSA768 in the RSA factoring challenge has been completed, and 1024 is not far from that.

Reply:
Indeed, whether to implement this feature and with which appropriate best practices require some discussions that are outside of the scope of this PR. We can open an issue to work on this.

Reply:
@atombrella I didn't mean the 1024 lower limit, I meant the *modulo* 1024 which isn't present in the current certbot code.

Reply:
Indeed I was also talking about the modulo, which would need a proper migration path for users that are not using standard key sizes.

Sadly also we need to revert back to the `assert` control on the key size. Indeed `assert` is operating only during the debug, so technically certbot is not supposed to fail at that point if you ask for a 512bits RSA key (even if it is bad and I am almost sure Boulder an any other decent CA server will refuse it).

But definitely we should discuss that outside of the PR and try to infer good practices through certbot, as it is done usually.

Reply:
OK. I removed the `% 1024` part. I don't really understand why `assert` is a better choice than the error. Personally, I prefer explicit error handling, rather than having `assert` statements tossed in. `import this` has "Explicit is better than implicit."

Reply:
To be clear: I also think an error is better than assert. However, since it was an assert before this PR, the code will break only during development, not on user side. If we change that to a proper exception, the code will break in all situations, which changes certbot behavior. In particular if some people are using short rsa keys, certbot would start to break unexpectedly.

I think we should definitely incite to use decent keys length, by refusing short keys in Certbot. However this needs a proper migration path.